test/common/upstream/load_stats_reporter_test.cc

```diff
@@ -221,6 +222,121 @@ TEST_F(LoadStatsReporterTest, ExistingClusters) {
   response_timer_cb_();
 }
 
+HostSharedPtr makeTestHost(const std::string& hostname,
+                           const ::envoy::config::core::v3::Locality& locality) {
+  const auto host = std::make_shared<NiceMock<::Envoy::Upstream::MockHost>>();
+  ON_CALL(*host, hostname()).WillByDefault(::testing::ReturnRef(hostname));
+  ON_CALL(*host, locality()).WillByDefault(::testing::ReturnRef(locality));
+  return host;
+}
+
+void addStats(const HostSharedPtr& host, double a, double b = 0, double c = 0, double d = 0) {
+  host->stats().rq_success_.inc();
+  host->loadMetricStats().add("metric_a", a);
+  if (b != 0) {
+    host->loadMetricStats().add("metric_b", b);
+  }
+  if (c != 0) {
+    host->loadMetricStats().add("metric_c", c);
+  }
+  if (d != 0) {
+    host->loadMetricStats().add("metric_d", d);
+  }
+}
+
+void addStatExpectation(envoy::config::endpoint::v3::UpstreamLocalityStats* stats,
+                        const std::string& metric_name, int num_requests_with_metric,
+                        double total_metric_value) {
+  auto metric = stats->add_load_metric_stats();
+  metric->set_metric_name(metric_name);
+  metric->set_num_requests_finished_with_metric(num_requests_with_metric);
+  metric->set_total_metric_value(total_metric_value);
+}
+
+// Validate that per-locality metrics are aggregated across hosts and included in the load report.
+TEST_F(LoadStatsReporterTest, UpstreamLocalityStats) {
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  expectSendMessage({});
+  createLoadStatsReporter();
+  time_system_.setMonotonicTime(std::chrono::microseconds(3));
+
+  // Set up some load metrics
+  NiceMock<MockClusterMockPrioritySet> cluster;
+  MockHostSet& host_set_ = *cluster.prioritySet().getMockHostSet(0);
+
+  ::envoy::config::core::v3::Locality locality0, locality1;
+  locality0.set_region("mars");
+  locality1.set_region("jupiter");
+  HostSharedPtr host0 = makeTestHost("host0", locality0), host1 = makeTestHost("host1", locality0),
+                host2 = makeTestHost("host2", locality1);
+  host_set_.hosts_per_locality_ = makeHostsPerLocality({{host0, host1}, {host2}});
+
+  addStats(host0, 0.11111, 1.0);
+  addStats(host0, 0.33333, 0, 3.14159);
+  addStats(host1, 0.44444, 0.12345);
+  addStats(host2, 10.01, 0, 20.02, 30.03);
+
+  cluster.info_->eds_service_name_ = "bar";
+  MockClusterManager::ClusterInfoMaps cluster_info{{{"foo", cluster}}, {}};
+  ON_CALL(cm_, clusters()).WillByDefault(Return(cluster_info));
+  deliverLoadStatsResponse({"foo"});
+  // First stats report on timer tick.
+  time_system_.setMonotonicTime(std::chrono::microseconds(4));
```

Comment:
consider using advanceTimeWait with a delta, rather than just setting the time to an absolute time.